File: tools/buildext/buildext.php
<?php

// make ext.go files for all exts

$dh = opendir('ext');
if (!$dh) die("could not open dir\n");

process_ext('core', 'Core', 'ext-core.go');

while(($ext = readdir($dh)) !== false) {
	if (($ext == '.') || ($ext == '..')) continue;

	process_ext('ext/'.$ext, $ext);
}

function process_ext($path, $ext, $output = 'ext.go') {
	echo $ext.' ('.$path.')'."\n";

	$constants = [];
	$functions = [];
	$classes = [];

	$it = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($path));

	foreach($it as $f => $finfo) {
		if (($f == '.') || ($f == '..')) continue;
		if ($f == $output) continue; // skip
		if (substr($f, -3) != '.go') continue; // skip non .go files
		if (!$finfo->isFile()) continue;

		$fp = fopen($f, 'r');
		if (!$fp) die("failed to open $f\n");

		$lineno = 0;
		$function_pending = NULL;
		$package = NULL;

		while(!feof($fp)) {
			$lineno += 1;
			$lin = fgets($fp);
			if (substr($lin, 0, 8) == 'package ') {
				// we're going to assume no comment after this
				$package = trim(substr($lin, 8));
				continue;
			}
			if ((substr($lin, 0, 5) == 'func ') && (!is_null($function_pending))) {
				// we have a function name
				$lin = substr($lin, 5);
				$pos = strpos($lin, '(');
				if ($pos===false) die("failed to parse function line func $lin\n");
				$func_name = substr($lin, 0, $pos);
				$functions[$function_pending]['val'] = $func_name;
				$function_pending = NULL;
				continue;
			}
			if ((substr($lin, 0, 5) != '// > ') && (substr($lin, 0, 4) != '//> ')) continue;
			$lin = trim(substr($lin, 4));
			$pos = strpos($lin, ' ');
			if ($pos === false) die("failed to parse $lin\n");
			$code = substr($lin, 0, $pos);
			$lin = trim(substr($lin, $pos+1));

			switch($code) {
				case 'const':
					// $lin is : <code> <value> [ // possible comment]
					$pos = strpos($lin, '//');
					if ($pos !== false) {
						$lin = trim(substr($lin, 0, $pos));
					}
					$pos = strpos($lin, ':');
					if ($pos === false) {
						die("failed to parse const $lin (no :)\n");
					}
					$const = trim(substr($lin, 0, $pos));
					$val = trim(substr($lin, $pos+1));
					$constants[$const] = ['pkg' => $package, 'val' => $val, 'where' => $f.':'.$lineno];
					break;
				case 'func':
					// $lin is: <return_type> <function_name> ( <arguments> )
					$pos = strpos($lin, ' ');
					if ($pos === false) {
						die("failed to parse func $lin (no space)\n");
					}
					$type = trim(substr($lin, 0, $pos));
					$lin = trim(substr($lin, $pos+1));

					$pos = strpos($lin, ' ');
					if ($pos === false) {
						die("failed to parse func $lin (no space)\n");
					}
					$func = trim(substr($lin, 0, $pos));
					$lin = trim(substr($lin, $pos+1));

					// TODO args
					$functions[strtolower($func)] = ['pkg' => $package, 'val' => null, 'where' => $f.':'.$lineno];
					$function_pending = strtolower($func);
					break;
				case 'class':
					// lin is: <class name> (should be a variable with this name)
					$pos = strpos($lin, '//');
					if ($pos !== false) {
						$lin = trim(substr($lin, 0, $pos));
					}
					$classes[$lin] = ['pkg' => $package, 'class' => $lin, 'where' => $f.':'.$lineno];
					break;
				default:
					die("failed to parse $code $lin (unknown code)\n");
			}
		}
		fclose($fp);
		if (!is_null($function_pending)) die("failed to find implementation of $function_pending");
	}

	$fp = fopen($path.'/'.$output.'~', 'w');
	if ($ext == 'Core') {
		fwrite($fp, 'package core'."\n\n");
		$prefix = '';
	} else {
		fwrite($fp, 'package '.$ext."\n\n");
		$prefix = 'core.';
	}
	fwrite($fp, "import \"github.com/MagicalTux/goro/core\"\n\n"); // other imports will be handled automatically at build time
	fwrite($fp, "// WARNING: This file is auto-generated. DO NOT EDIT\n\n");
	fwrite($fp, "func init() {\n");
	fwrite($fp, "\tphpctx.RegisterExt(&phpctx.Ext{\n");
	fwrite($fp, "\t\tName: \"".addslashes($ext)."\",\n"); // addslashes not quite equivalent to go's %q
	fwrite($fp, "\t\tVersion: ".$prefix."VERSION,\n");

	fwrite($fp, "\t\tClasses: []phpv.ZClass{\n");
	ksort($classes);
	foreach($classes as $class => $info) {
		if ($info['pkg'] != $ext) $class = $info['pkg'].'.'.$class;
		fwrite($fp, "\t\t\t".$class.",\n");
	}
	fwrite($fp, "\t\t},\n");

	fwrite($fp, "\t\tFunctions: map[string]*phpctx.ExtFunction{\n");
	ksort($functions);
	foreach($functions as $func => $info) {
		// sample args: Args: []*core.ExtFunctionArg{&core.ExtFunctionArg{ArgName: "output"}, &core.ExtFunctionArg{ArgName: "...", Optional: true}}
		fwrite($fp, "\t\t\t\"".addslashes($func)."\": &phpctx.ExtFunction{Func: ".$info['val'].", Args: []*phpctx.ExtFunctionArg{}},\n"); // TODO args
	}
	fwrite($fp, "\t\t},\n");

	fwrite($fp, "\t\tConstants: map[phpv.ZString]phpv.Val{\n");
	ksort($constants);
	foreach($constants as $const => $info) {
		fwrite($fp, "\t\t\t\"".addslashes($const)."\": ".$info['val'].",\n");
	}
	fwrite($fp, "\t\t},\n");
	fwrite($fp, "\t})\n");
	fwrite($fp, "}\n");
	fclose($fp);

	// rename
	rename($path.'/'.$output.'~', $path.'/'.$output);
}

File: tools/buildext/buildext.go
package main

import (
	"bytes"
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"io"
	"io/fs"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

type IdentiferMap struct {
	GoIdent  string
	PhpIdent string
	Package  string
	IsAlias  bool
}

type ExtData struct {
	Dirname   string
	Imports   map[string]string
	Functions map[string]IdentiferMap
	Constants map[string]IdentiferMap
	Classes   map[string]IdentiferMap
}

func main() {
	type ExtEntry struct {
		dirname  string
		destfile string
	}

	entries := []ExtEntry{
		{dirname: "core", destfile: "ext-core.go"},
	}
	files, err := os.ReadDir("ext")
	if err != nil {
		panic(err)
	}
	for _, f := range files {
		if !f.IsDir() {
			continue
		}
		entries = append(entries, ExtEntry{
			dirname:  path.Join("ext", f.Name()),
			destfile: "ext.go",
		})
	}

	for _, entry := range entries {
		ext, err := processExt(entry.dirname, entry.destfile)
		if err != nil {
			panic(err)
		}
		destfile := filepath.Join(entry.dirname, entry.destfile)
		file, err := os.OpenFile(destfile, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
		if err != nil {
			panic(err)
		}
		println("generated", destfile)
		writeExt(ext, file)
		file.Close()
	}

}

func writeExt(ext *ExtData, w io.Writer) {
	var constants []string
	var funcNames []string
	var classes []string
	for k := range ext.Constants {
		constants = append(constants, k)
	}
	for k := range ext.Functions {
		funcNames = append(funcNames, k)
	}
	for k := range ext.Classes {
		classes = append(classes, k)
	}

	sort.Strings(funcNames)
	sort.Strings(constants)
	sort.Strings(classes)

	template := `
package %s

import (%s)

// WARNING: This file is auto-generated. DO NOT EDIT

func init() {
	phpctx.RegisterExt(&phpctx.Ext{
		Name:    "%s",
		Version: %s,
		Classes: []*phpobj.ZClass{%s},
		// Note: ExtFunctionArg is currently unused
		Functions: map[string]*phpctx.ExtFunction{%s},
		Constants: map[phpv.ZString]phpv.Val{%s},
	})
}
	`
	pkg := path.Base(ext.Dirname)
	extName := pkg
	version := "VERSION"
	importSet := map[string]struct{}{
		// "github.com/MagicalTux/goro/core":        {},
		"github.com/MagicalTux/goro/core/phpctx": {},
		"github.com/MagicalTux/goro/core/phpobj": {},
		"github.com/MagicalTux/goro/core/phpv":   {},
	}

	if ext.Dirname != "core" {
		importSet["github.com/MagicalTux/goro/core"] = struct{}{}
		version = "core.VERSION"
	} else {
		extName = "Core"
	}

	var buf bytes.Buffer
	var importStr string
	var functionStr string
	var constantStr string
	var classStr string

	if len(ext.Functions) > 0 {
		buf.WriteRune('\n')
	}

	maxLen := 0
	for _, phpIdent := range funcNames {
		maxLen = max(maxLen, len(phpIdent))
	}

	for _, phpIdent := range funcNames {
		decl := ext.Functions[phpIdent]
		if decl.Package != ext.Dirname {
			pkg := "github.com/MagicalTux/goro/" + decl.Package
			importSet[pkg] = struct{}{}
		}

		goIdent := decl.GoIdent
		if decl.Package != ext.Dirname {
			goIdent = path.Base(decl.Package) + "." + goIdent
		}

		comment := ""
		if decl.IsAlias {
			comment = " // alias"
		}
		indent := strings.Repeat(" ", maxLen-len(phpIdent)+1)
		format := "\t\t\t" + `"%s":%s{Func: %s, Args: []*phpctx.ExtFunctionArg{}},%s` + "\n"
		buf.WriteString(fmt.Sprintf(format, phpIdent, indent, goIdent, comment))
	}
	if len(ext.Functions) > 0 {
		buf.WriteString("\t\t")
	}
	functionStr = buf.String()
	buf.Reset()

	if len(ext.Classes) > 0 {
		buf.WriteRune('\n')
	}
	for _, phpIdent := range classes {
		decl := ext.Classes[phpIdent]
		if decl.Package != ext.Dirname {
			pkg := "github.com/MagicalTux/goro/" + decl.Package
			importSet[pkg] = struct{}{}
		}

		goIdent := decl.GoIdent
		if decl.Package != ext.Dirname {
			goIdent = path.Base(decl.Package) + "." + goIdent
		}

		format := "\t\t\t" + `%s,` + "\n"
		buf.WriteString(fmt.Sprintf(format, goIdent))
	}
	if len(ext.Classes) > 0 {
		buf.WriteString("\t\t")
	}
	classStr = buf.String()
	buf.Reset()

	if len(ext.Constants) > 0 {
		buf.WriteRune('\n')
	}
	maxLen = 0
	for _, constant := range constants {
		maxLen = max(maxLen, len(constant))
	}
	for _, constant := range constants {
		decl := ext.Constants[constant]
		if decl.Package != ext.Dirname {
			pkg := "github.com/MagicalTux/goro/" + decl.Package
			importSet[pkg] = struct{}{}
		}

		goIdent := decl.GoIdent
		if decl.Package != ext.Dirname {
			goIdent = path.Base(decl.Package) + "." + goIdent
		}

		indent := strings.Repeat(" ", maxLen-len(constant)+1)
		format := "\t\t\t" + `"%s":%s%s,` + "\n"
		buf.WriteString(fmt.Sprintf(format, constant, indent, goIdent))
	}
	if len(ext.Constants) > 0 {
		buf.WriteString("\t\t")
	}
	constantStr = buf.String()
	buf.Reset()

	imports := []string{}
	for k := range importSet {
		imports = append(imports, k)
	}

	sort.Strings(imports)

	if len(importSet) > 0 {
		buf.WriteRune('\n')
	}
	for _, imp := range imports {
		buf.WriteString("\t\"")
		buf.WriteString(imp)
		buf.WriteString("\"\n")
	}
	importStr = buf.String()
	buf.Reset()

	output := fmt.Sprintf(template, pkg, importStr, extName, version, classStr, functionStr, constantStr)
	output = strings.TrimSpace(output) + "\n"
	w.Write([]byte(output))
}

func processExt(dirname, destFilename string) (*ExtData, error) {
	ext := &ExtData{
		Dirname:   dirname,
		Imports:   make(map[string]string),
		Functions: make(map[string]IdentiferMap),
		Constants: make(map[string]IdentiferMap),
		Classes:   make(map[string]IdentiferMap),
	}
	err := filepath.WalkDir(dirname, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if path == filepath.Join(dirname, destFilename) {
			return nil
		}
		if filepath.Ext(path) == ".go" {
			processExtFile(path, ext)
		}
		return nil
	})

	if err != nil {
		return nil, err
	}

	return ext, nil
}

func processExtFile(filename string, ext *ExtData) {
	fset := token.NewFileSet()
	f, err := parser.ParseFile(fset, filename, nil, parser.ParseComments)
	if err != nil {
		fmt.Println(err)
		return
	}
	pkg := path.Dir(filename)

	consts := getPhpConstants(f, pkg)
	funcs := getPhpFunctions(f, pkg)
	classes := getPhpClasses(f, pkg)

	for k, id := range consts {
		ext.Constants[k] = id
	}
	for k, id := range funcs {
		ext.Functions[k] = id
	}
	for k, id := range classes {
		ext.Classes[k] = id
	}
}

func parseCommentDirective(s string) []string {
	if !strings.HasPrefix(s, "//") {
		return nil
	}
	s = strings.TrimLeft(s[2:], " \t")
	if len(s) == 0 {
		return nil
	}
	if s[0] != '>' {
		return nil
	}
	s = strings.TrimLeft(s[1:], " \t")

	re := regexp.MustCompile(`\s+`)
	return re.Split(s, -1)
}

func parseFunctionSignature(s string) string {
	re := regexp.MustCompile(`\/\/\s*>\s*func\s*[\w|]+\s*(\w*)\s*\(.*\)`)
	matches := re.FindStringSubmatch(s)
	if len(matches) >= 1 {
		return matches[1]
	}
	return ""
}
func parseClassSignature(s string) (string, bool) {
	re := regexp.MustCompile(`\/\/\s*>\s*class\s*(\w*)\s*`)
	matches := re.FindStringSubmatch(s)
	if matches == nil {
		return "", false
	}

	var name string
	if len(matches) >= 1 {
		name = matches[1]
	}
	return name, true
}

func parseConstantSignature(s string) *string {
	re := regexp.MustCompile(`\/\/\s*>\s*const\s*(\w*)\s*`)
	matches := re.FindStringSubmatch(s)
	if matches == nil {
		return nil
	}

	var name string
	if len(matches) >= 1 {
		name = matches[1]
	}
	return &name
}

func getConstantSignature(commentGroup *ast.CommentGroup) *string {
	if commentGroup == nil {
		return nil
	}
	for _, comment := range commentGroup.List {
		text := parseConstantSignature(comment.Text)
		if text != nil {
			return text
		}
	}
	return nil
}

func getPhpFunctions(f *ast.File, pkg string) map[string]IdentiferMap {
	result := map[string]IdentiferMap{}
	for _, d := range f.Decls {
		switch decl := d.(type) {
		case *ast.FuncDecl:
			if decl.Doc == nil {
				continue
			}

			phpIdent := ""
			nonWord := regexp.MustCompile(`[^a-zA-Z_0-9]`)
			var aliases []string

			for _, comment := range decl.Doc.List {
				fields := parseCommentDirective(comment.Text)
				if len(fields) == 0 {
					continue
				}
				switch fields[0] {
				case "func":
					if len(fields) < 3 {
						panic(fmt.Errorf("invalid func comment: %s", comment.Text))
					}

					phpIdent = fields[2]
					if nonWord.MatchString(phpIdent) {
						panic(fmt.Errorf("invalid func name: %s", phpIdent))
					}

				case "alias":
					aliases = append(aliases, fields[1:]...)
				}
			}

			if phpIdent != "" {
				if _, ok := result[phpIdent]; ok {
					panic(fmt.Errorf("function name is already used: %s", phpIdent))
				}

				result[phpIdent] = IdentiferMap{
					GoIdent:  decl.Name.Name,
					PhpIdent: phpIdent,
					Package:  pkg,
				}

				for _, alias := range aliases {
					if _, ok := result[alias]; ok {
						panic(fmt.Errorf("function name is already used: %s", alias))
					}
					result[alias] = IdentiferMap{
						GoIdent:  decl.Name.Name,
						PhpIdent: alias,
						Package:  pkg,
						IsAlias:  true,
					}
				}
			}

		}
	}

	return result
}

func getPhpClasses(f *ast.File, pkg string) map[string]IdentiferMap {
	result := map[string]IdentiferMap{}
	for _, decl := range f.Decls {
		decl, ok := decl.(*ast.GenDecl)
		if !ok {
			continue
		}
		for i, spec := range decl.Specs {
			spec, ok := spec.(*ast.ValueSpec)
			if !ok || len(spec.Names) == 0 {
				continue
			}

			var phpIdent string
			ok = false

			if spec.Doc != nil {
				phpIdent, ok = parseClassSignature(spec.Doc.List[0].Text)
			}
			if !ok && i == 0 && decl.Doc != nil {
				phpIdent, ok = parseClassSignature(decl.Doc.List[0].Text)
			}

			if ok && phpIdent == "" {
				phpIdent = spec.Names[0].Name
			}

			if phpIdent != "" {
				result[phpIdent] = IdentiferMap{
					PhpIdent: phpIdent,
					GoIdent:  spec.Names[0].Name,
					Package:  pkg,
				}
			}
		}
	}
	return result
}

func getPhpConstants(f *ast.File, pkg string) map[string]IdentiferMap {
	result := map[string]IdentiferMap{}
	for _, decl := range f.Decls {

		decl, ok := decl.(*ast.GenDecl)
		if !ok {
			continue
		}

		declDoc := getConstantSignature(decl.Doc)

		for _, spec := range decl.Specs {
			spec, ok := spec.(*ast.ValueSpec)
			if !ok || len(spec.Names) == 0 {
				continue
			}

			for i, id := range spec.Names {
				specDoc := getConstantSignature(spec.Doc)

				if declDoc == nil && specDoc == nil {
					continue
				}

				if specDoc == nil {
					result[id.Name] = IdentiferMap{
						GoIdent:  id.Name,
						PhpIdent: id.Name,
						Package:  pkg,
					}
					continue
				}

				phpIdent := id.Name
				if *specDoc != "" && i == 0 {
					phpIdent = *specDoc
				}

				result[phpIdent] = IdentiferMap{
					GoIdent:  id.Name,
					PhpIdent: phpIdent,
					Package:  pkg,
				}
			}
		}
	}
	return result
}
